modules/common/src/main/scala/notebook/front/widgets/charts/Chart.scala

```diff
@@ -0,0 +1,232 @@
+  extends DataToRenderableConverter[C](originalData, maxPoints)
+  with JsWorld[Seq[(String, Any)], Seq[(String, Any)]]
+  with Texts
+  with Utils {
+  import notebook.JSBus._
+
+  def sizes:(Int, Int)=(600, 400)
+
+  @volatile var currentC = originalData
+  @volatile var currentPoints = initialItems
+  @volatile var currentMax = maxPoints
+
+  def approxTotalItemCount(): String = {
+    sampler.samplingStrategy match {
+      // on DataFrames, do not call df.count() as it's rather expensive
+      case magic.LimitBasedSampling() =>
+        val sampledCount = currentPoints.length
+        if (currentMax > sampledCount) s"$sampledCount" else s"$sampledCount or more"
+      case _ => s"${toPoints.count(currentC)}"
+    }
+  }
+
+  def samplingWarningMsg(): String = {
+    sampler.samplingStrategy match {
+        case magic.LimitBasedSampling() =>
+          if (currentMax > currentPoints.length) ""
+          else " (Warning: showing only first " + currentMax + " rows)"
+
+        case _ if currentMax <= toPoints.count(currentC) =>
+          " (Warning: randomly sampled "+currentMax + " entries)"
+
+        case _ => ""
+    }
+  }
+
+  // initialize sampling warning on Chart initialization
+  val totalRowCount = outWithInitialValue(approxTotalItemCount)
+  val warnSamplingInUse = outWithInitialValue(samplingWarningMsg)
+
+  // ---- Helpers to mutate the chart reactively ----
+  // ------------------------------------------------
+  def updateChartStatus() = {
+    warnSamplingInUse(samplingWarningMsg)
+    totalRowCount(approxTotalItemCount)
+  }
+
+  def newMax(max:Int) = {
+    //update state
+    currentMax = max
+    applyOn(currentC)
+  }
+
+  def applyOn(newData:C) = apply {
+    currentC = newData
+    currentPoints = toPoints(newData, currentMax)
+    val d = currentPoints map mToSeq
+    updateChartStatus()
+    this.apply(d)
+    d
+  }
+
+  //val log = org.slf4j.LoggerFactory.getLogger("Chart")
+  private[this] var first = true
+  def addAndApply(otherData:C, resetInit:Boolean=false) = {
+    if (resetInit && first) {
+      first = false
+      applyOn(otherData)
+    } else {
+      apply {
+        currentC = toPoints.append(currentC, otherData)
+        currentPoints = toPoints(currentC, currentMax)
+        updateChartStatus()
+        val d = currentPoints map mToSeq
+        this.apply(d)
+        d
+      }
+    }
+  }
+
+  override val singleCodec = jsStringAnyCodec
+  override val singleToO = identity[Seq[(String, Any)]] _
+
+  val extendedContent:Option[scala.xml.Elem] = None
+
+  override val content = Some {
+    val container = <div>
+      <span class="chart-total-item-count">{totalRowCount.toHtml} entries total</span>
+      <span class="chart-sampling-warning" style="color:red">{warnSamplingInUse.toHtml}</span>
+      <div>
+      </div>
+    </div>
+    extendedContent.map(c => container.copy(child = container.child ++ c)).getOrElse(container)
+  }
+}
+
+trait Sequencifiable[C] { self: Chart[C] =>
```

Comment:
this is wtf name. only after huge refactoring, I could understand that `Seq[(String, Any)]` is used only inside Javascript.

let's come up with a different name. 
also why is it needed at all? i.e. why  should it be different per chart?
can't we have it shared among all charts, inside `DataToRenderableConverter`?